lib/active_shipping/external_return_label_request.rb

```diff
@@ -0,0 +1,478 @@
+# -*- coding: utf-8 -*-
+module ActiveShipping
+
+  class ExternalReturnLabelRequest
+
+    CAP_STRING_LEN = 100
+
+    USPS_EMAIL_REGEX = /^([a-zA-Z0-9_\-\.]+)@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.)|(([a-zA-Z0-9\-]+\.)+))([a-zA-Z]{2,4}|[0-9]{1,3})(\]?)$/
+
+    LABEL_FORMAT = {
+      'Instructions' => 'null',
+      'No Instructions' => 'NOI',
+      'Double Label' => 'TWO'
+    }
+
+    SERVICE_TYPE_CODE = [
+      '044', '019', '596', '020', '597','022', '024', '017', '018'
+    ]
+
+    CALL_CENTER_OR_SELF_SERVICE = ['CallCenter', 'Customer']
+
+    LABEL_DEFINITION = ['4X6', 'Zebra-4X6', '4X4', '3X6']
+
+    IMAGE_TYPE = ['PDF', 'TIF']
+
+    attr_reader :customer_name,
+                :customer_address1,
+                :customer_address2,
+                :customer_city,
+                :customer_state,
+                :customer_zipcode,
+                :customer_urbanization,
+                :company_name,
+                :attention,
+                :label_format,
+                :label_definition,
+                :service_type_code,
+                :merchandise_description,
+                :insurance_amount,
+                :address_override_notification,
+                :packaging_information,
+                :packaging_information2,
+                :call_center_or_self_service,
+                :image_type,
+                :address_validation,
+                :sender_name,
+                :sender_email,
+                :recipient_name,
+                :recipient_email,
+                :recipient_bcc,
+                :merchant_account_id,
+                :mid
+
+    def initialize(options = {})
+      options.each do |pair|
+        self.public_send("#{pair[0]}=".to_sym, pair[1]) if self.respond_to?("#{pair[0]}=".to_sym)
+      end
+
+      verify_or_raise_required
+    end
+
+    def self.from_hash(options={})
+      self.new(options)
+    end
+
+    def to_bool(v, default = false)
+      v = v.to_s
+      if v =~ (/(true|yes|1)$/i)
+        true
+      elsif v =~ (/(false|no|0)$/i)
+        false
+      else
+        default
+      end
+    end
+
+    def sanitize(v)
+      if v.is_a?(String)
+        v.strip!
+        v[0..CAP_STRING_LEN - 1]
+      else
+        nil
+      end
+    end
+
+    # Sent by the system containing the returns label attachment and message.
+    def recipient_bcc=(v)
```

Comment:
I love the level of validation in this, so much friendlier than the rest of Active Shipping where you get back some obscure error message that has nothing to do with what you did wrong. Great work finding out and coding the requirements for all of these fields. :+1: 

And don't even get me started on the doc strings describing what each field actually means, miles beyond the opaque tag names everywhere else in the codebase. So many times I've had to track down postal service docs just to figure out what a tag is. :heart_eyes_cat: 

Lots of copy-paste though. I think these could all be refactored to use some private `validate_x` methods. I think `validate_email(s)`, `validate_string_length(s,max_len)`, `validate_range(n,min,max)`, `validate_included(v,options)` that test a value and then raise a good error would cover everything and then each method could just be like
```ruby
def recipient_email=(v)
  v = sanitize(v)
  validate_email(v)
  @recipient_email = v
end
```